In [9]:
import PyPDF2
import jieba
import jieba.posseg as pseg
from jieba import analyse
import numpy as np
import os

# 1. Import Text from pdf

In [2]:
#create file object variable
#opening method will be rb
Lin=open('Lin.pdf','rb')
Yu=open('Yu.pdf','rb')

# 2. Data Exploration

In [3]:
with open("Lin.pdf", "rb") as pdf_file:
    read_pdf = PyPDF2.PdfFileReader(pdf_file)
    number_of_pages = read_pdf.getNumPages()
    page = read_pdf.pages[0]
    page_content = page.extractText()
print("論文共: ",number_of_pages,"頁")
print(page_content)

論文共:  96 頁
		doi:10.6342/NTU201700203
 
 國立臺灣大學社會科學院國家發展研究所  
碩士論文  
Graduate Institute of National Development 
College of Social Sciences 
National Taiwan University 
Master Thesis  
 
三人競選 之中槓桿者的政治社會基礎及其影響  
以2014 新竹市長選舉為例  
The Political and Social Basis of the Role of Leverage 
in a Three Person Campaign 
-An  Analysis of the 2014 Hsinchu City Mayoral 
Election  
 
林智堅  
Chih-Chien Lin 
 
指導教授：陳明通  博士 
Advisor: Ming-Tong Chen Ph.D 
   
中華民國 106 年1月 


In [4]:
with open("Lin.pdf", "rb") as pdf_file:
    read_pdf = PyPDF2.PdfFileReader(pdf_file)
    number_of_pages = read_pdf.getNumPages()
    page = read_pdf.pages[10]
    page_content = page.extractText()
print(page_content)

		doi:10.6342/NTU201700203
10 
 五、 根據上述的投票取向，如何進一步分析槓桿者的槓桿作用 ？ 
第三節  研究設計  
壹、研究途徑與研究方 法 
首先，所謂方法論（ methodology ）就是關於研究過程的分析工作，或對方法
及研究法的批判工作。方法（ method ）只是收集及處理資料的技術；研究途徑
（approach ）卻是選擇問題及資料的準則，因此方法（ method ）和研究途徑
（approach ）有著本質上的不同6。 
研究選舉有許多不同的途徑， 學者黃智聰、程小綾從經濟景氣的角度切入7，
分析臺灣 縣市長選舉與政黨輪替之間的關聯性，而學者鄧志松選擇從 選舉地理學
的概念切入，分析臺灣 三次總統大選8，不過依據學者 蔡佳泓、王金壽、王鼎銘的
整理，過去 研究地方層級的選舉，大致上採取以下三種研究途徑9，第一個是個別
縣市選民的投票行為； 第二則是描述地方派系在 選舉過程中動員選民的方式及成
效；第三是從分立政府的角度切入，探討探討選民的投票行為。  
而根據學者何金銘的整理， 美國關於研究選民投票行為的途徑主要有以下三
種研究途徑10，第一個是社會學研究途徑（ sociological approach ），主要探討選民
的宗教、社會經濟及區位等因素，對選民投票行為的影響；第二個則是 社會心理
學研究途徑（ social psychological approach ），主要從政黨認同、政見取向及候選
人取向，探討對選民投票行為的影響；第三個則是 經濟學途徑（ economic 
approach ），假設選民是理性且有充裕的資訊，選民會經過理性的計算，進而選出                                                                                                                            5  胡佛(1998 ) 。《政治學的科學探究 (三)政治參與與選舉行為》 ，臺北市：三民書局。頁 177 -213 。 6  易君博 (2003 ) 。《政治理論與研究方法》。臺北：三民書局，頁 97。 7  黃智聰、程小綾  ，2005 ，〈經濟投票與政黨輪替 --以臺灣縣市長選舉為例〉，《選舉研究》，
第十

In [5]:
with open("Yu.pdf", "rb") as pdf_file:
    read_pdf = PyPDF2.PdfFileReader(pdf_file)
    number_of_pages = read_pdf.getNumPages()
    page = read_pdf.pages[0]
    page_content = page.extractText()
print("論文共: ",number_of_pages,"頁")
print(page_content)

論文共:  97 頁
doi:10.6342/NTU201601231
國立臺灣大學社會科學院國家發展研究所 碩士論文 
Graduate Institute of National Development  College of Social Sciences
 National Taiwan University Master Thesis 
 2014年新竹市長選舉研究： 林智堅勝選的政治社會基礎
 
A Study of Hsinchu City Mayor Election: Social-Political Basis of Election Winner Lin Chih-Chien  余正煌 Cheng-Huang Yu  指導教授：李碧涵博士 Advisor: Bih-Hearn Lee, Ph.D.  中華民國105年7月 July, 2016  
  


In [6]:
with open("Yu.pdf", "rb") as pdf_file:
    read_pdf = PyPDF2.PdfFileReader(pdf_file)
    number_of_pages = read_pdf.getNumPages()
    page = read_pdf.pages[10]
    page_content = page.extractText()
print(page_content)

doi:10.6342/NTU201601231
如前述，現任新竹市長林智堅在一路不被看好的情形下當選勝出，本研究更深究其中造成林智堅勝選更深層的社會、政治，甚至個人因素。扣除掉前述對外公開之民調結果，從更深一層的新竹市選民結構來看，新竹市選民結構，深藍選民占23%，淺藍選民占21%；深綠占21%，淺綠占8%，另有27%中間選民，整體而言，選民結構上國民黨（藍營）較具優勢，藍綠選民比率約在六比四左右。
7根 據中央選舉委員會公布過去10年新竹市總統、立法委員及市長選舉結果，有效票數得票率統計如次： 一、2012年區域立法委員選舉，國民黨53.27%、民進黨41.85%； 二、2009年縣市長選舉，國民黨55.63%、民進黨41.32%； 三、2008年總統選舉，國民黨64.7%、民進黨35.3%； 四、2008年區域立法委員選舉，國民黨60.62%、民進黨38.12%； 五、2005年市長選舉，國民黨69.27%、民進黨30.73%； 總體來說，國民黨得票率最低仍有 53.27%以上，基本盤皆相當穩定。而民進黨得票率則在30.73%至41.85%之間遊移。因此，從過往選舉結果數據統計及基本選民結構來看2014年選舉結果，新竹市長當選人怎麼看都不可能是林智堅。有鑑於此，林智堅勝選已非「表面上」之民調結果、選舉結果數據統計及基本選民結構等因素所能夠解釋，本研究將以更細微的角度，從林智堅所具備之政治、社會基礎及立基於選民的投票取向切入，逐一疏理分析林智堅之勝選因素。 
第二節  研究目的與主要研究問題  
壹、研究目的  2014年新竹市長選舉結果，民進黨籍候選人林智堅得到 76,578張選票，獲票率達38.35%，當選民選新竹市市長；國民黨許明財得票數為 75,564，獲票率為 37.85%； 無黨籍蔡仁堅得票40,480票 ， 獲票比例為20.27%； 無黨籍劉正幸得票1,914 票，獲票比例為0.95%；無黨籍吳淑敏得票5,101，獲票比例2.55%。
8在民主政治
7參見未來事件交易所，http://www.xsmart4.com/ploy/ploy_detail.asp?id=1080&ClassID=677&KeywordID=all&SearchString_Data=&page=1&lastpage=10，檢索日期2015年12月28日。 
8中央選委會，

# 3. Data Preprocess

In [56]:
text=[]
with open("Lin.pdf", "rb") as pdf_file:
    read_pdf = PyPDF2.PdfFileReader(pdf_file)
    number_of_pages = read_pdf.getNumPages()
    for i in range(number_of_pages-1):
        page = read_pdf.pages[i]
        page_content = page.extractText()
        text.append(page_content)
        
str_L=' '
for t in text:
        str_L+=t+' '
        
text_file = open("Lin.txt", "w", encoding='UTF-8')
text_file.write(str_L)
text_file.close()

In [57]:
text=[]
with open("Yu.pdf", "rb") as pdf_file:
    read_pdf = PyPDF2.PdfFileReader(pdf_file)
    number_of_pages = read_pdf.getNumPages()
    for i in range(number_of_pages-1):
        page = read_pdf.pages[i]
        page_content = page.extractText()
        text.append(page_content)
        
str_Y=' '
for t in text:
        str_Y+=t+' '
        
text_file = open("Yu.txt", "w", encoding='UTF-8')
text_file.write(str_Y)
text_file.close()

# 4. Import Data

In [12]:
file_a='Lin.txt';file_b='Yu.txt'
str_a = '';str_b = ''
if not os.path.isfile(file_a):
    print(file_a,"is not file")
elif not os.path.isfile(file_b):
    print(file_b,"is not file")
else:
    with open(file_a,'r', encoding='UTF-8') as f:
        for line in f.readlines():
            str_a += line.strip()
                
        f.close()
    with open(file_b,'r', encoding='UTF-8') as f:
        for line in f.readlines():
            str_b += line.strip()
                
        f.close()

# 5. 分詞

In [13]:
def splitWords(str_a):
    wordsa=pseg.cut(str_a)
    cuta = ""
    seta = set()
    for key in wordsa:
        #print(key.word,key.flag)
        cuta += key.word + " "
        seta.add(key.word) 
    return [cuta, seta]

# 6. Jaccard Similarity
ref: https://medium.com/data-science-bootcamp/understand-jaccard-index-jaccard-similarity-in-minutes-25a703fbf9d7

In [14]:
def JaccardSim(str_a,str_b):
    '''
    len（sa & sb）/ len（sa | sb）
    '''
    seta = splitWords(str_a)[1]
    setb = splitWords(str_b)[1]
    sa_sb = 1.0 * len(seta & setb) / len(seta | setb)
    return sa_sb

In [32]:
print("林志堅跟余正煌的論文相似度 (Jaccard Simialrity): ", round(JaccardSim(str_a,str_b)*100,2),"%")

林志堅跟余正煌的論文相似度 (Jaccard Simialrity):  38.19 %


# 7. Cosine Similarity
ref: https://medium.com/geekculture/cosine-similarity-and-cosine-distance-48eed889a5c4

In [24]:
def cos_sim(a,b):
    a = np.array(a)
    b = np.array(b)
    return np.sum(a*b)/(np.sqrt(np.sum(a**2))*np.sqrt(np.sum(b**2)))

## Term Frequency(TF) — Inverse Dense Frequency(IDF)
ref: https://medium.com/analytics-vidhya/tf-idf-term-frequency-technique-easiest-explanation-for-text-classification-in-nlp-with-code-8ca3912e58c3

In [28]:
def countIDF(text,topK):
    tfidf = analyse.extract_tags
    cipin = {}
    fenci = jieba.cut(text)
    for word in fenci:
        if word not in cipin.keys():
            cipin[word] = 0
        cipin[word] += 1
    keywords = tfidf(text,topK,withWeight=True)
    ans = []
    for keyword in keywords:
        ans.append(cipin[keyword[0]])
    return ans

In [34]:
def splitWord_CosineSimlaryty(str_a,str_b,topK = 20,sim =cos_sim):
    vec_a = countIDF(str_a,topK)
    vec_b = countIDF(str_b,topK)
    return sim(vec_a,vec_b)

In [35]:
print("林志堅跟余正煌的論文相似度 (Cosine Simialrity): ", round(splitWord_CosineSimlaryty(str_a,str_b)*100,2),"%")

林志堅跟余正煌的論文相似度 (Cosine Simialrity):  93.56 %


# 8 Persion Similarity
ref: https://medium.com/@cavaldovinos/human-pose-estimation-pose-similarity-dc8bf9f78556

In [36]:
def pers_sim(a,b):
    a = np.array(a)
    b = np.array(b)
    a = a - np.average(a)
    b = b - np.average(b)
    return np.sum(a*b) / (np.sqrt(np.sum(a**2))*np.sqrt(np.sum(b**2)))

In [37]:
def splitWord_PersionSimlaryty(str_a,str_b,topK = 20,sim =pers_sim):
    vec_a = countIDF(str_a,topK)
    vec_b = countIDF(str_b,topK)
    return sim(vec_a,vec_b)

In [38]:
print("林志堅跟余正煌的論文相似度 (Cosine Simialrity): ", round(splitWord_PersionSimlaryty(str_a,str_b)*100,2),"%")

林志堅跟余正煌的論文相似度 (Cosine Simialrity):  52.26 %


# Summary

| Methodology | 林志堅跟余正煌的論文相似度 |
| :--: | :--: |
| Jaccard Similarity | 38.19% |
| Cosine Similarity | 93.56% |
| Persion Similarity | 52.26% |